### PR TITLE
introduce SourceAnalyzer API for ides

### PR DESCRIPTION
**Pull request by matklad**

### CODE REVIEW

**Comment by flodiebold** [comment]
Looks like an improvement, though I'm not a fan of the name ;)

File: crates/ra_ide_api/src/call_info.rs [hunk 1]
```diff
@@ -18,19 +17,26 @@ pub(crate) fn call_info(db: &RootDatabase, position: FilePosition) -> Option<Cal
...
     let name_ref = calling_node.name_ref()?;
 
-    // Resolve the function's NameRef (NOTE: this isn't entirely accurate).
-    let file_symbols = crate::symbol_index::index_resolve(db, name_ref);
-    let symbol = file_symbols.into_iter().find(|it| it.ptr.kind() == FN_DEF)?;
-    let fn_file = db.parse(symbol.file_id);
-    let fn_def = symbol.ptr.to_node(&fn_file);
-    let fn_def = ast::FnDef::cast(fn_def).unwrap();
-    let function = hir::source_binder::function_from_source(db, symbol.file_id, fn_def)?;
+    let analyser = hir::SourceAnalyzer::new(db, position.file_id, name_ref.syntax());
```
Comment (kjeremy) [inline crates/ra_ide_api/src/call_info.rs]: Should be `analyzer`

**Comment by kjeremy** [comment]
This is a really nice cleanup. One note: I think it's a little weird that there are `resolve_*` methods and a separate `resolver()` that are exposed.

**Comment by matklad** [comment]
@kjeremy yeah, I am not really satisfied with the current state either, because it seems like there's a lot of duplication between SourceAnalyzer, variouse specifc methods of source_binder and Resolver. Hope to takle this soon!

**Comment by flodiebold** [comment]
:+1: for the new name!

I guess this could probably also replace the `ScopesWithSourceMap` API...

**Comment by matklad** [comment]
Ok, I think this is in "better than status quo" state now, what do you think about merging it as is and hiding more details behind SourceAnalyzer front-end in follow-up PRs?

File: crates/ra_ide_api/src/completion/complete_dot.rs [hunk 2]
```diff
@@ -312,6 +305,30 @@ mod tests {
         kind: Method,
         detail: "pub fn blah(&self)"
     }
+]"###
+        );
+    }
+
+    #[test]
+    fn test_completion_works_in_consts() {
```
Comment (flodiebold) [inline crates/ra_ide_api/src/completion/complete_dot.rs]: :tada:

**Pull request merged**